independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java

```diff
@@ -339,6 +341,12 @@ private void notifyObservers(T event, ObserverExceptionHandler exceptionHandler,
                 Predicate<ObserverMethod<? super T>> predicate) {
             EventContext eventContext = new EventContextImpl<>(event, eventMetadata);
             for (ObserverMethod<? super T> observerMethod : observerMethods) {
+                Bean<?> declaringBean = observerMethod.getDeclaringBean();
```

Comment:
Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

Reply:
> Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

It is, there are failures otherwise; note that this code is not here for reception handling.
Rather it is there to handle cases where an observer belongs to a bean with context that's inactive - so for instance req. context bean with observer but req. context wasn't activated. Originally, Arc was still trying to invoke it ending with `ContextNotActiveException`.
The dependent and `null` cases are for dependent beans which have no context of their own and for synth. observers which can belong to no bean.

Reply:
I see and I wonder if the spec acutally covers this use case. In any case, I would rather move the logic to the generated `ObserverMethod#notify()` so that it's all in one place.

Reply:
> I see and I wonder if the spec acutally covers this use case

One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

> I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.

Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

Reply:
> > I see and I wonder if the spec acutally covers this use case
> 
> One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

Oook, it probably makes sense. Is it worth a clarification issue?

> 
> > I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.
> 
> Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

Logically yes, but from maintainability POV it's better to change this [line](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java#L417)  to `notify.ifNull(context).trueBranch().returnVoid();` :shrug:

Reply:
This is already in [spec](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#observer_notification):

> If there is no context active for the scope to which the bean declaring the observer method belongs, then the observer method should not be called.

> Logically yes, but from maintainability POV it's better...

Fair enough.